cookiecutter/generate.py

```diff
@@ -313,13 +304,14 @@ def generate_files(
     :param keep_project_on_failure: If `True` keep generated project directory even when
         generation fails
     """
-    template_dir = find_template(repo_dir)
-    logger.debug('Generating project from %s...', template_dir)
     context = context or OrderedDict([])
```

Comment:
Nice one

It should work fine since the line below `context = context or OrderedDict([])` will provide the default value in case context comes in as `None`.

That said, I would personally prefer this default value to live in the parameter definition, it would be clearer and avoid confusion
However I'm not sure if we really can do so, if any function/caller were to call this with an explicit `context=None` then the default would not set it to the OrderedDict and it will indeed fail. We would need to check each place this is called, and if this `generate_files` function is considered part of the public api (is there one?) I guess it would be a breaking change

With the above in mind I would say this refactor would be great, but is out-of-scope for this PR

Reply:
Function parameters should generally never have mutable objects as their defaults, because each time the function is called, that default parameter is never re-instantiated.

```python
def test(a=[]):
  a.append(1)
  print(a)

test()  # [1]
test()  # [1, 1]
```

The current design -- `None` as the default, and getting transformed into a brand new `OrderedDict([])` -- follows best practices. Let's consider this discussion resolved. :+1:

Reply:
Oh 🤦 missed this and that is very true, all the more reason to leave it aside for now. Thanks!